### PR TITLE
Add handler that provides some backoff for accepts if these failed be…

### PR DESCRIPTION
**Pull request by normanmaurer**
…cause of an IOError.

Motivation:

Often accept errors are recoverable over time as for example you may just run out of file descriptors. Because of this it may be useful to just "delay" accepts a bit to recover.

Modifications:

Add ChannelHandler which will backoff accepts.

Result:

Be able to recover from accept errors gracefully.

### CODE REVIEW

**Comment by normanmaurer** [comment]
This is the second part of https://github.com/apple/swift-nio/pull/125... You can add this `ChannelHandler` to the `ChannelPipeline` of the `ServerChannel` to implement some sort of backoff  on accept errors.

I still need to write some tests tho, just wanted to share first

**Comment by normanmaurer** [comment]
also @toffaletti

**Comment by weissi** [comment]
@normanmaurer needs rebase

File: Sources/NIO/ChannelHandlers.swift [hunk 1]
```diff
@@ -17,6 +17,80 @@
...
+    }
+    
+    public func handlerRemoved(ctx: ChannelHandlerContext) {
+        /// If the channel is not active anymore we should cancel the previous scheduled read operation.
+        if !ctx.channel.isActive, let scheduled = scheduledRead {
+            scheduled.cancel()
+        }
+    }
+    
+    private func scheduleRead(_ in: TimeAmount, _ ctx: ChannelHandlerContext) {
```
Comment (weissi) [inline Sources/NIO/ChannelHandlers.swift]: these arguments should be labelled I think

File: Sources/NIO/ChannelHandlers.swift [hunk 2]
```diff
@@ -17,6 +17,80 @@
 
 import struct Dispatch.DispatchTime
 
+
+/// `ChannelHandler` which implements some sort of backoff for a `ServerChannel` once `accept` will produce an `IOError`. This is useful as often these errors are recoverable in the sense that we can recover if we slow down the accept rate a bit.
+public final class AcceptBackoffHandler: ChannelDuplexHandler {
```
Comment (weissi) [inline Sources/NIO/ChannelHandlers.swift]: can we have a few tests for that?

File: Sources/NIO/ChannelHandlers.swift [hunk 3]
```diff
@@ -17,6 +17,80 @@
...
+    ///
+    /// - parameters:
+    ///     - backoffProvider: returns a `TimeAmount` to backoff more `reads` (a.k.a accepts).
+    public init(backoffProvider: @escaping (IOError) -> TimeAmount?) {
+        self.backoffProvider = backoffProvider
+    }
+    
+    public func read(ctx: ChannelHandlerContext) {
+        // If we already have a read scheduled there is no need to schedule another one.
+        if scheduledRead == nil {
```
Comment (weissi) [inline Sources/NIO/ChannelHandlers.swift]: `guard scheduledRead == nil else { return }`?

File: Sources/NIO/ChannelHandlers.swift [hunk 4]
```diff
@@ -17,6 +17,80 @@
...
+            if let backoffTime = currentBackoffTime {
+                let now = Int(DispatchTime.now().uptimeNanoseconds)
+                if now >= backoffTime {
+                    // The backoff already expired, just do a read.
+                    self.doRead(ctx)
+                } else {
+                    // Schedule the read to be executed after the backoff time elapsed.
+                    scheduleRead(TimeAmount.nanoseconds(backoffTime - now), ctx)
```
Comment (weissi) [inline Sources/NIO/ChannelHandlers.swift]: here you schedule a read with a time in nanoseconds _relative_ to `now`

File: Sources/NIO/ChannelHandlers.swift [hunk 5]
```diff
@@ -17,6 +17,80 @@
...
+        }
+    }
+    
+    public func errorCaught(ctx: ChannelHandlerContext, error: Error) {
+        if let ioError = error as? IOError {
+            if let amount = backoffProvider(ioError) {
+                currentBackoffTime = Int(DispatchTime.now().uptimeNanoseconds) + amount.nanoseconds
+                if let scheduled = scheduledRead {
+                    scheduled.cancel()
+                    scheduleRead(TimeAmount.nanoseconds(currentBackoffTime!), ctx)
```
Comment (weissi) [inline Sources/NIO/ChannelHandlers.swift]: here you schedule a read with a time in nanoseconds _absolute_ (including `now`). So something can't be right here.

Comment (weissi) [reply Sources/NIO/ChannelHandlers.swift]: `TimeInterval` is supposed to be relative to now not an absolute amount of time.

File: Sources/NIO/ChannelHandlers.swift [hunk 6]
```diff
@@ -17,6 +17,80 @@
 
 import struct Dispatch.DispatchTime
 
+
+/// `ChannelHandler` which implements some sort of backoff for a `ServerChannel` once `accept` will produce an `IOError`. This is useful as often these errors are recoverable in the sense that we can recover if we slow down the accept rate a bit.
+public final class AcceptBackoffHandler: ChannelDuplexHandler {
+    public typealias InboundIn = Channel
+    public typealias OutboundIn = Channel
+    
+    private var currentBackoffTime: Int?
```
Comment (weissi) [inline Sources/NIO/ChannelHandlers.swift]: this looks like it is an absolute time in nano seconds since boot of the machine, maybe we could improve the name? `nextReadAbsoluteTimeNS` isn't pretty but maybe more clear?

**Comment by normanmaurer** [comment]
Will rework this one as soon as https://github.com/apple/swift-nio/pull/125 is in. Then I can also write some tests ;)

**Comment by normanmaurer** [comment]
@weissi @Lukasa tests added

File: Sources/NIO/ChannelHandlers.swift [hunk 7]
```diff
@@ -17,6 +17,88 @@
 
 import struct Dispatch.DispatchTime
 
+
+/// `ChannelHandler` which implements some sort of backoff for a `ServerChannel` once `accept` will produce an `IOError`. This is useful as often these errors are recoverable in the sense that we can recover if we slow down the accept rate a bit.
```
Comment (Lukasa) [inline Sources/NIO/ChannelHandlers.swift]: This needs rewriting:

> A `ChannelHandler` that implements a backoff for a `ServerChannel` when `accept` produces an `IOError`. These errors are often recoverable by reducing the rate at which we call accept.

File: Sources/NIO/ChannelHandlers.swift [hunk 8]
```diff
@@ -17,6 +17,88 @@
...
+    
+    private var nextReadDeadlineNS: Int?
+    private let backoffProvider: (IOError) -> TimeAmount?
+    private var scheduledRead: Scheduled<Void>?
+    
+    /// Create a new instance
+    ///
+    /// - parameters:
+    ///     - backoffProvider: returns a `TimeAmount` to backoff more `reads` (a.k.a accepts).
+    public init(backoffProvider: @escaping (IOError) -> TimeAmount?) {
```
Comment (Lukasa) [inline Sources/NIO/ChannelHandlers.swift]: Should this really be implemented in the form of a closure? At the very least shouldn't we provide a default "reasonable estimate" implementation that people can plug in, rather than require them to know how this works?

File: Sources/NIO/ChannelHandlers.swift [hunk 9]
```diff
@@ -17,6 +17,88 @@
...
+            scheduled.cancel()
+            self.scheduledRead = nil
+        }
+        self.nextReadDeadlineNS = nil
+        ctx.fireChannelInactive()
+    }
+    
+    public func handlerRemoved(ctx: ChannelHandlerContext) {
+        /// If the channel is not active anymore we should cancel the previous scheduled read operation.
+        if !ctx.channel.isActive, let scheduled = self.scheduledRead {
```
Comment (Lukasa) [inline Sources/NIO/ChannelHandlers.swift]: The channel being active here is not relevant: if we are removed from the channel we *must* cancel the scheduled read, because the scheduled read uses our `ChannelHandlerContext`, which is no longer in the pipeline. 

I think the way this check should actually go is that if the channel is active when this handler is removed, and there's a scheduled read, that read should happen *now*.

File: Tests/NIOTests/AcceptBackoffHandlerTest.swift [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+import XCTest
+@testable import NIO
+import Dispatch
+import NIOConcurrencyHelpers
+
+
+public class AcceptBackoffHandlerTest : XCTestCase {
```
Comment (Lukasa) [inline Tests/NIOTests/AcceptBackoffHandlerTest.swift]: No space before the colon.

File: Tests/NIOTests/AcceptBackoffHandlerTest.swift [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+                readCount += 1
+                ctx.read()
+            }
+        }
+
+        let group = MultiThreadedEventLoopGroup(numThreads: 1)
+        defer {
+            XCTAssertNoThrow(try group.syncShutdownGracefully())
+        }
+        class NonAcceptingServerSocket : ServerSocket {
```
Comment (Lukasa) [inline Tests/NIOTests/AcceptBackoffHandlerTest.swift]: No space before the colon.

File: Sources/NIO/ChannelHandlers.swift (hunk 9 above)
Comment (Lukasa) [reply Sources/NIO/ChannelHandlers.swift]: More generally, this code branch isn't tested. Can you test it?

File: Sources/NIO/ChannelHandlers.swift [hunk 12]
```diff
@@ -17,6 +17,88 @@
...
+        ctx.fireErrorCaught(error)
+    }
+    
+    public func channelInactive(ctx: ChannelHandlerContext) {
+        if let scheduled = self.scheduledRead {
+            scheduled.cancel()
+            self.scheduledRead = nil
+        }
+        self.nextReadDeadlineNS = nil
+        ctx.fireChannelInactive()
```
Comment (Lukasa) [inline Sources/NIO/ChannelHandlers.swift]: This code branch isn't tested: can you test it?

File: Sources/NIO/ChannelHandlers.swift [hunk 13]
```diff
@@ -17,6 +17,88 @@
...
+    
+    public func errorCaught(ctx: ChannelHandlerContext, error: Error) {
+        if let ioError = error as? IOError {
+            if let amount = backoffProvider(ioError) {
+
+                self.nextReadDeadlineNS = Int(DispatchTime.now().uptimeNanoseconds) + amount.nanoseconds
+                if let scheduled = self.scheduledRead {
+                    scheduled.cancel()
+                    scheduleRead(in: amount, ctx: ctx)
+                }
```
Comment (Lukasa) [inline Sources/NIO/ChannelHandlers.swift]: This code branch isn't tested, can you test it?

File: Sources/NIO/ChannelHandlers.swift [hunk 14]
```diff
@@ -17,6 +17,88 @@
...
+    ///
+    /// - parameters:
+    ///     - backoffProvider: returns a `TimeAmount` to backoff more `reads` (a.k.a accepts).
+    public init(backoffProvider: @escaping (IOError) -> TimeAmount?) {
+        self.backoffProvider = backoffProvider
+    }
+    
+    public func read(ctx: ChannelHandlerContext) {
+        // If we already have a read scheduled there is no need to schedule another one.
+        guard scheduledRead == nil else { return }
```
Comment (Lukasa) [inline Sources/NIO/ChannelHandlers.swift]: This early return isn't tested either, can you test it?

File: Sources/NIO/ChannelHandlers.swift (hunk 9 above)
Comment (normanmaurer) [reply Sources/NIO/ChannelHandlers.swift]: @Lukasa good point... that said I think we need to trigger a read as otherwise we may end-up never reading again. Let me just triggered the read directly in this case. And yep will add more tests.

File: Sources/NIO/ChannelHandlers.swift (hunk 8 above)
Comment (normanmaurer) [reply Sources/NIO/ChannelHandlers.swift]: @Lukasa I think the closure thing is nice... maybe we should use a default implementation that use 1 second tho.

**Comment by normanmaurer** [comment]
@Lukasa adding more tests.

File: Sources/NIO/ChannelHandlers.swift (hunk 12 above)
Comment (normanmaurer) [reply Sources/NIO/ChannelHandlers.swift]: done

File: Sources/NIO/ChannelHandlers.swift (hunk 14 above)
Comment (normanmaurer) [reply Sources/NIO/ChannelHandlers.swift]: done

**Comment by normanmaurer** [comment]
@Lukasa @weissi tests added...

File: Sources/NIO/ChannelHandlers.swift [hunk 15]
```diff
@@ -17,6 +17,94 @@
 
 import struct Dispatch.DispatchTime
 
+
+/// A `ChannelHandler` that implements a backoff for a `ServerChannel` when accept produces an IOError. `
```
Comment (Lukasa) [inline Sources/NIO/ChannelHandlers.swift]: You've got a backtick flying out at the end of your sentence here.

File: Sources/NIO/ChannelHandlers.swift [hunk 16]
```diff
@@ -17,6 +17,94 @@
...
+/// These errors are often recoverable by reducing the rate at which we call accept.
+public final class AcceptBackoffHandler: ChannelDuplexHandler {
+    public typealias InboundIn = Channel
+    public typealias OutboundIn = Channel
+    
+    private var nextReadDeadlineNS: Int?
+    private let backoffProvider: (IOError) -> TimeAmount?
+    private var scheduledRead: Scheduled<Void>?
+
+    /// Default implementation used as `backoffProvider` which backoff for 1 second.
```
Comment (Lukasa) [inline Sources/NIO/ChannelHandlers.swift]: s/which backoff for/which delays `accept` by/

File: Sources/NIO/ChannelHandlers.swift [hunk 17]
```diff
@@ -17,6 +17,94 @@
...
+    public static func defaultBackoffProvider(error: IOError) -> TimeAmount? {
+        return .seconds(1)
+    }
+
+    /// Create a new instance
+    ///
+    /// - parameters:
+    ///     - backoffProvider: returns a `TimeAmount` to backoff more `reads` (a.k.a accepts).
```
Comment (Lukasa) [inline Sources/NIO/ChannelHandlers.swift]: Let's reword this to: "returns a `TimeAmount` which will be the amount of time to wait before attempting another `read`"

File: Sources/NIO/ChannelHandlers.swift (hunk 13 above)
Comment (normanmaurer) [reply Sources/NIO/ChannelHandlers.swift]: done

File: Sources/NIO/ChannelHandlers.swift [hunk 18]
```diff
@@ -17,6 +17,94 @@
...
+
+    /// Default implementation used as `backoffProvider` which delays accept by 1 second.
+    public static func defaultBackoffProvider(error: IOError) -> TimeAmount? {
+        return .seconds(1)
+    }
+
+    /// Create a new instance
+    ///
+    /// - parameters:
+    ///     - backoffProvider: returns a `TimeAmount` which will be the amount of time to wait before attempting another `read.
```
Comment (Lukasa) [inline Sources/NIO/ChannelHandlers.swift]: Unbalanced backtick.

**Comment by normanmaurer** [comment]
I really have no idea why this always happen in XCode :(

File: Sources/NIO/ChannelHandlers.swift (hunk 18 above)
Comment (normanmaurer) [reply Sources/NIO/ChannelHandlers.swift]: and fixed.... sorry for it
